namerd/storage/zk/src/main/scala/com/twitter/finagle/serverset2/buoyant/ZkSession.scala

```diff
@@ -0,0 +1,186 @@
+
+    // Upon initial connection, send auth info, then update `u`
+    newClient.state.changes.filter {
+      _ == WatchState.SessionState(SessionState.SyncConnected)
+    }.toFuture.unit.before {
+      // TODO: send auth info
+      Future.Unit
+    }.onSuccess { _ =>
+      logger.info(s"New ZKSession is connected. Session ID: ${sessionId(newClient)}")
+      client() = newClient
+      reconnectBackoff.reset()
+    }
+
+    // Kick off a delayed reconnection on session expiration.
+    newClient.state.changes.filter {
+      _ == WatchState.SessionState(SessionState.Expired)
+    }.toFuture().unit.before {
+      val jitter = reconnectBackoff.next()
+      logger.error(s"Zookeeper session ${sessionId(newClient)} has expired. Reconnecting in $jitter")
+      Future.sleep(jitter)
+    }.ensure { reconnect() }
+  }
+
+  // If the zookeeper cluster is under duress, there can be 100's of thousands of clients
+  // attempting to read and write at once. Limit to a (fairly large) concurrent request cap.
+  // Use a semaphore (versus explicit rate limiting) to approximate the throughput of the cluster.
+  // N.B. this semaphore has no max-waiters limit. This could lead to an OOME if the zk operations
+  // never complete. This is preferable to handling and re-queuing (via future.sleep etc)
+  // the error if an arbitrary max-limit is set.
+  private[this] val limiter = new AsyncSemaphore(100)
+  private[this] val waitersGauge = statsReceiver.addGauge("numWaiters") { limiter.numWaiters }
+
+  private[this] def limit[T](f: => Future[T]): Future[T] =
+    limiter.acquire().flatMap { permit =>
+      f.ensure {
+        // don't release the permit until f is complete
+        permit.release()
+      }
+    }
+
+  private[this] def retryWithDelay[T](f: => Future[T]): Future[T] =
+    Future.sleep(retryBackoff.next()).before(f)
+
+  /**
+   * Invoke a `Future[T]`-producing operation, retrying on
+   * [[com.twitter.finagle.serverset2.client.KeeperException.ConnectionLoss]]
+   * according to a backoff schedule defined by [[retryBackoff]]. The operation itself
+   * will be limited by the session-level semaphore.
+   */
+  private def safeRetry[T](go: => Future[T]): Future[T] = {
+    def loop(): Future[T] =
+      limit { go }.rescue {
+        case exc: KeeperException.ConnectionLoss =>
+          logger.warning(s"ConnectionLoss to Zookeeper host. Session $sessionId. Retrying")
+          retryWithDelay { loop() }
+      }
+
+    loop()
+  }
+
+  /**
+   * A persistent operation: reissue a watched operation every
+   * time the watch fires, applying safe retries when possible.
+   *
+   * The returned Activity is asynchronous: watches aren't reissued
+   * when the Activity is no longer observed.
+   */
+  def watchedOperation[T](go: => Future[Watched[T]]): Activity[T] =
+    Activity(Var.async[Activity.State[T]](Activity.Pending) { u =>
+      @volatile var closed = false
+
+      def loop(): Future[Unit] = {
```

Comment:
It's not clear to me why this returns Future[Unit] and not just Unit.  Perhaps just to satisfy retryWithDelay's signature?